fix(remix-dev/vite): keep code-split JS files in SSR build

hi-ogawa opened the pull request:
<!--

👋 Hey, thanks for your interest in contributing to Remix!

If this is a simple docs change, go ahead and delete all this.

**Please ask first before starting work on any significant new features.**

It's never a fun experience to have your pull request declined after investing a
lot of time and effort into a new feature. To avoid this from happening, we
request that contributors create a
[Feature Request discussion](https://github.com/remix-run/remix/discussions/new?category=ideas)
to first discuss any significant new features.

https://github.com/remix-run/remix/blob/main/CONTRIBUTING.md

Please fill in or delete each item below:

-->

Closes: https://github.com/remix-run/remix/issues/8041
Follow up for https://github.com/remix-run/remix/pull/7892

EDIT: following comment is for initial approach. Now it's written to remove only unnecessary files based on the suggestion in  https://github.com/remix-run/remix/pull/8042#issuecomment-1817261780

---

Instead of keeping whole `assets` directory, maybe we could probe vite manifest and keep only `isDynamicEntry`.
(EDIT: I further extended reproduction https://github.com/hi-ogawa/remix-reproduction-8041 and I found that `isDynamicEntry` is not a right way to detect this)

For example, vite manifest from ssr build for `vite-build-test.ts` looks like this:

<details><summary>reveal build/.vite/manifest.json</summary>

(here `assets/test2-LxNzCzR4.txt` is server-only assets, so it's moved to client assets directory. `assets/test1-eT52RzBS.txt` and `assets/_virtual_server-entry-119NVDiu.css` are unnecessary files to keep.)

```json
{
  "app/assets/test1.txt": {
    "file": "assets/test1-eT52RzBS.txt",
    "src": "app/assets/test1.txt"
  },
  "app/assets/test2.txt": {
    "file": "assets/test2-LxNzCzR4.txt",
    "src": "app/assets/test2.txt"
  },
  "app/ssr-code-split-lib.ts": {
    "file": "assets/ssr-code-split-lib-l8BW9uKL.js",
    "isDynamicEntry": true,
    "src": "app/ssr-code-split-lib.ts"
  },
  "virtual:server-entry": {
    "assets": [
      "assets/test1-eT52RzBS.txt",
      "assets/test2-LxNzCzR4.txt"
    ],
    "css": [
      "assets/_virtual_server-entry-119NVDiu.css"
    ],
    "dynamicImports": [
      "app/ssr-code-split-lib.ts"
    ],
    "file": "index.js",
    "isEntry": true,
    "src": "virtual:server-entry"
  }
}
```

</details>

However, I don't think it's so important to clean up the rest so strictly (and also I'm not entirely sure if there could be more edge cases which I'm not aware of). So, I chose to keep the directory as is.
Please let me know what you think!

hi-ogawa commented on packages/remix-dev/vite/plugin.ts:
```diff
@@ -789,7 +793,7 @@ export const remixVitePlugin: RemixVitePlugin = (options = {}) => {
               let src = path.join(serverBuildDir, ssrAssetPath);
               let dest = path.join(assetsBuildDirectory, ssrAssetPath);
-              await fse.move(src, dest);
+              await fse.copy(src, dest);
```
For consistency, I also changed from `move` to `copy` so that ssr build assets will be kept intact.

markdalgleish commented:
@hi-ogawa Thanks for looking into this!

> I don't think it's so important to clean up the rest

I'd prefer for us to aim for a clean server build output so that people aren't deploying a tonne of unused assets along with their server code. Especially when we're still marked as unstable and have a bit of leeway to get this right.

Maybe a more safe approach, rather than removing the entire assets directory, is to do a more targeted removal of all `assets` and `css` files in the server manifest?

hi-ogawa commented:
> I'd prefer for us to aim for a clean server build output so that people aren't deploying a tonne of unused assets along with their server code. Especially when we're still marked as unstable and have a bit of leeway to get this right.

Thanks for the feedback!
Right, it makes sense to still experiment to strive for better while it's unstable.

The another concern I'm thinking  about is that, for experienced vite users (who are familiar with `manifest: true` and `ssrEmitAssets: true`) or other vite plugins making use of the same options, it might be simply unexpected for some assets to disappear which vite would usually emit.

This might be out of scope right now, but for example, for whatever reasons, such users (or other plugins) might wish to customize `rollupOptions.input/output` during ssr build to include more server entry files, which then might utilize `manifest.json` and assets files, which they might not want for remix to delete.
So, deleting things could take away the flexibility of vite plugin system, which I think is a strong aspect of vite ecosystem.

But, I totally understand that, for the most of use cases, this won't be an issue and remix doesn't want to bloat default server output which users are likely consume and deploy directly. Also probably removing `chunk.assets` and `chunk.css` files won't hurt anything, so I'll try that approach.

markdalgleish commented:
You're raising fair concerns for sure. I think we aim for a clean build output for now, but we can always add an option to disable this behaviour if it turns out it's a legitimate issue for someone.

hi-ogawa commented:
I updated a PR to cleanup unnecessary `chunk.assets` and `chunk.css` from server `assets` directory and also added a test case to verify it.
I would appreciate next round of reviews. Thanks!
